src/Aspire.Hosting/ProjectResourceBuilderExtensions.cs

```diff
@@ -620,6 +620,55 @@ public static IResourceBuilder<ProjectResource> WithEndpointsInEnvironment(
         return builder;
     }
 
+    /// <summary>
+    /// <remarks>
+    /// When the executable resource is converted to a container resource, the arguments to the executable
+    /// are not used. This is because arguments to the project often contain physical paths that are not valid
+    /// in the container. The container can be set up with the correct arguments using the <paramref name="configure"/> action.
+    /// </remarks>
+    /// Adds support for containerizating this <see cref="ProjectResource"/> during deployment.
+    /// The resulting container image is built, and when the optional <paramref name="configure"/> action is provided,
+    /// it is used to configure the container resource.
+    /// </summary>
+    /// <typeparam name="T">Type of executable resource</typeparam>
+    /// <param name="builder">Resource builder</param>
+    /// <param name="configure">Optional action to configure the container resource</param>
+    /// <returns>A reference to the <see cref="IResourceBuilder{T}"/>.</returns>
+    public static IResourceBuilder<T> PublishAsDockerFile<T>(this IResourceBuilder<T> builder, Action<IResourceBuilder<ContainerResource>>? configure = null)
+        where T : ProjectResource
+    {
+        if (!builder.ApplicationBuilder.ExecutionContext.IsPublishMode)
+        {
+            return builder;
+        }
+
+        // The implementation here is less than ideal, but we don't have a clean way of building resource types
+        // that change their behavior based on the context. In this case, we want to change the behavior of the
+        // resource from an ProjectResource to a ContainerResource. We do this by removing the ProjectResource
+        // from the application model and adding a new ContainerResource in its place in publish mode.
+
+        // There are still dangling references to the original ProjectResource in the application model, but
+        // in publish mode, it won't be used. This is a limitation of the current design.
+        builder.ApplicationBuilder.Resources.Remove(builder.Resource);
+
+        var container = new ProjectContainerResource(builder.Resource);
+        var cb = builder.ApplicationBuilder.AddResource(container);
+        // WithImage makes this a container resource (adding the annotation)
+        cb.WithImage(builder.Resource.Name);
+        cb.WithDockerfile(contextPath: builder.Resource.GetProjectMetadata().ProjectPath);
+        // Arguments to the executable often contain physical paths that are not valid in the container
+        // Clear them out so that the container can be set up with the correct arguments
+        cb.WithArgs(c => c.Args.Clear());
+
+        configure?.Invoke(cb);
+
+        // Even through we're adding a ContainerResource
+        // update the manifest publishing callback on the original ProjectResource
+        // so that the container resource is written to the manifest
+        return builder.WithManifestPublishingCallback(context =>
```

Comment:
How does the original project's manifest publishing callback get invoked? We removed it from the ApplicationBuilder.

Reply:
The reference is in user code, this is just a fail safe.